src/Common/src/System/Text/ValueStringBuilder.cs

```diff
@@ -0,0 +1,189 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text
+{
+    internal ref struct ValueStringBuilder
+    {
+        private char[] _arrayToReturnToPool;
+        private Span<char> _chars;
+        private int _pos;
+
+        public ValueStringBuilder(Span<char> initialBuffer)
+        {
+            _arrayToReturnToPool = null;
+            _chars = initialBuffer;
+            _pos = 0;
+        }
+
+        public string GetString()
+        {
+            var s = new string(_chars.Slice(0, _pos));
+
+            char[] toReturn = _arrayToReturnToPool;
+            this = default; // for safety, to avoid using pooled array if this instance is erroneously appended to again
+
+            if (toReturn != null)
+            {
+                ArrayPool<char>.Shared.Return(toReturn);
+            }
+
+            return s;
+        }
+
+        public int Length => _pos;
+
+        public bool TryCopyTo(Span<char> destination, out int charsWritten)
+        {
+            if (_pos > destination.Length)
```

Comment:
could be just
```c#
 +            if (! _chars.Slice(0, _pos).TryCopyTo(destination))
 +            {
 +                charsWritten = 0;
 +                return false;
               }
             
```
as span repeats the check ?

Reply:
I'll change it.  The current approach slightly benefits the case where the destination isn't long enough, whereas your suggestion slightly benefits the case where the destination is long enough, and presumably the latter is more common.